### PR TITLE
Warn if paper_trail_on_destroy(:after) is combined with ActiveRecord belongs_to_required_by_default

### PR DESCRIPTION
**Pull request by md5**
Based on 18d35ef (see #683, #682)

cc/ @owenr

### CODE REVIEW

File: lib/paper_trail/has_paper_trail.rb [hunk 1]
```diff
@@ -115,14 +115,23 @@ def setup_callbacks_from_options(options_on = [])
...
-      def paper_trail_on_destroy(recording_order = 'after')
+      def paper_trail_on_destroy(recording_order = 'before')
         unless %w[after before].include?(recording_order.to_s)
           fail ArgumentError, 'recording order can only be "after" or "before"'
         end
 
-        send "#{recording_order}_destroy",
-             :record_destroy,
-             :if => :save_version?
+        if recording_order.to_s == 'after' and
```
Comment (md5) [inline lib/paper_trail/has_paper_trail.rb]: Note, I changed this from `recording_order == 'after'` to `recording_order.to_s == 'after'` to ensure that the warning is still issued when a the `:after` symbol is used. This fix should probably be made on `master` as well.

**Comment by md5** [comment]
Incidentally, I'm not sure I agree with the fix in #683. I probably would have set `required: false` on the `belongs_to` and added a conditional `validates_presence_of :item, unless: -> { event == 'destroy' }` (or some such).

**Comment by jaredbeck** [comment]
Hmm, it looks like CI didn't run, maybe because this PR isn't against the master branch?

**Comment by md5** [comment]
@jaredbeck I'm not sure. The only setting I'm familiar with in Travis that affects whether a branch builds is the "Build only if .travis.yml is present" setting.

**Comment by jaredbeck** [comment]
I can't find anything in their docs, so I reached out to Travis support. If you want to get CC'd on it, send me your email. Mine's jared@jaredbeck.com.

**Comment by md5** [comment]
@jaredbeck If you don't mind I can wait for a report back.

**Comment by owenr** [comment]
Re: changing to optional relationship, see discussion in #682.
Maybe CI didn't run because this was cherry-picked?

**Comment by md5** [comment]
Cherry picked commits get a new SHA, so Travis should build it based on the
PR.

As for changing the relationship to required, I don't see much discussion
in #682, other than @jaredbeck saying he'd like it to be required (which is
not the case in Rails 4, as I'm sure you both know)

**Comment by jaredbeck** [comment]
I got a response back from Travis CI:

> Thanks for reaching out and sorry for this issue.
>
> I had a look and it seems GitHub returned a 404 when trying to get the payload for the commit on this pull request. This is an edge case that is not handled well on our side at the moment. Sometimes it's caused by a recent change to permissions of one members/collaborators of your repository or organization. Hence, do you remember changing the permissions recently?
>
> Can you try pushing a new commit (it could be an empty one) on the PR to see if things are back on their feet?
>
> Thank you for your understanding and please keep us posted on the status of this issue. We will follow-up upon your reply.

Please try pushing a new commit, Mike.

File: lib/paper_trail/has_paper_trail.rb (hunk 1 above)
Comment (jaredbeck) [inline lib/paper_trail/has_paper_trail.rb]: Good idea! 

> This fix should probably be made on master as well.

Please do, thanks!

**Comment by md5** [comment]
@jaredbeck I did a no-op `--amend` and pushed a new commit. Looks like Travis is running now.

File: spec/models/callback_modifier_spec.rb [hunk 2]
```diff
@@ -29,7 +29,7 @@
           it 'should default to after destroy' do
```
Comment (jaredbeck) [inline spec/models/callback_modifier_spec.rb]: s/after/before/

**Comment by jaredbeck** [comment]
Do you understand this `Bundler could not find compatible versions for gem "activesupport"` error? I don't. o_O

File: lib/paper_trail/has_paper_trail.rb (hunk 1 above)
Comment (md5) [inline lib/paper_trail/has_paper_trail.rb]: I opened https://github.com/airblade/paper_trail/pull/813

I added a spec to test that the warning is actually being issued and it's not passing for some reason. We can discuss on the PR.

File: spec/models/callback_modifier_spec.rb (hunk 2 above)
Comment (md5) [inline spec/models/callback_modifier_spec.rb]: Not sure how I missed that while pulling in the cherry-picked commit... Fixed in 65c3ce4426d2da1fa51df8b3e38091cc6def685a

Comment (md5) [inline spec/models/callback_modifier_spec.rb]: And I pushed the version comparison fix I made in #813 as a new amended commit: cd4c667f5dfd306748b87c355f39d00bbb53ea86

**Comment by md5** [comment]
@jaredbeck I'm not sure what the deal is with that Bundler failure though.

**Comment by jaredbeck** [comment]
Mike, please rebase. The dependency resolution issue should be fixed by https://github.com/airblade/paper_trail/commit/29786c36b121fe8b0088737463fa24ed508d2e47

**Comment by md5** [comment]
@jaredbeck Rebased

**Comment by jaredbeck** [comment]
Mike, I just realized this is a breaking change. We're trying to follow semver, so we can't have a breaking change in 4.2. Would you mind if we just add the warning message, and do not change the default order? Sorry for realizing this so late, I guess I was focused on getting the travis build to pass.

**Comment by md5** [comment]
@jaredbeck Updated to just be a warning

**Comment by jaredbeck** [comment]
Thanks Mike! Sorry for all the grief with the Travis build.

**Comment by md5** [comment]
No worries

**Pull request merged**